pkg/job/scraper.go

```diff
@@ -0,0 +1,241 @@
+	// Setup so we only do one GetAccount call per region + role combo when running jobs
+	roleRegionToAccount := map[model.Role]map[string]func() (string, error){}
+	jobConfigVisitor(s.jobsCfg, func(_ any, role model.Role, region string) {
+		if _, exists := roleRegionToAccount[role]; !exists {
+			roleRegionToAccount[role] = map[string]func() (string, error){}
+		}
+		roleRegionToAccount[role][region] = sync.OnceValues[string, error](func() (string, error) {
+			accountID, err := s.runnerFactory.GetAccountClient(region, role).GetAccount(ctx)
+			if err != nil {
+				return "", fmt.Errorf("failed to get Account: %w", err)
+			}
+			return accountID, nil
+		})
+	})
+
+	var wg sync.WaitGroup
+	mux := &sync.Mutex{}
+	jobErrors := make([]Error, 0)
+	metricResults := make([]model.CloudwatchMetricResult, 0)
+	resourceResults := make([]model.TaggedResourceResult, 0)
+	s.logger.Debug("Starting job runs")
+
+	jobConfigVisitor(s.jobsCfg, func(job any, role model.Role, region string) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			var namespace string
+			jobAction(s.logger, job, func(job model.DiscoveryJob) {
+				namespace = job.Type
+			}, func(job model.CustomNamespaceJob) {
+				namespace = job.Namespace
+			})
+			jobContext := JobContext{
+				Namespace: namespace,
+				Region:    region,
+				RoleARN:   role.RoleArn,
+			}
+			jobLogger := s.logger.With("namespace", jobContext.Namespace, "region", jobContext.Region, "arn", jobContext.RoleARN)
+
+			accountID, err := roleRegionToAccount[role][region]()
+			if err != nil {
+				jobError := NewError(jobContext, "Account for job was not found", err)
+				mux.Lock()
+				jobErrors = append(jobErrors, jobError)
+				mux.Unlock()
+				return
+			}
+			jobContext.AccountID = accountID
+			jobLogger = jobLogger.With("account", jobContext.AccountID)
+
+			var jobToRun cloudwatchrunner.Job
+			jobAction(jobLogger, job,
+				func(job model.DiscoveryJob) {
+					jobLogger.Debug("Starting resource discovery")
+					rmRunner := s.runnerFactory.NewResourceMetadataRunner(jobLogger, region, role)
+					resources, err := rmRunner.Run(ctx, region, job)
+					if err != nil {
+						jobError := NewError(jobContext, "Failed to run resource metadata for job", err)
+						mux.Lock()
+						jobErrors = append(jobErrors, jobError)
+						mux.Unlock()
+
+						return
+					}
+					if len(resources) > 0 {
+						result := model.TaggedResourceResult{
+							Context: jobContext.ToScrapeContext(job.CustomTags),
+							Data:    resources,
+						}
+						mux.Lock()
+						resourceResults = append(resourceResults, result)
+						mux.Unlock()
+					} else {
+						jobLogger.Debug("No tagged resources")
+					}
+					jobLogger.Debug("Resource discovery finished", "number_of_discovered_resources", len(resources))
+
+					jobToRun = cloudwatchrunner.DiscoveryJob{Job: job, Resources: resources}
+				}, func(job model.CustomNamespaceJob) {
+					jobToRun = cloudwatchrunner.CustomNamespaceJob{Job: job}
+				},
+			)
```

Comment:
I wanted to keep resource discovery here and somewhat obvious. In the future when resource discovery happens elsewhere, I would expect this to be something like

```suggestion
			jobAction(jobLogger, job,
				func(job model.DiscoveryJob) {
					jobToRun = cloudwatchrunner.DiscoveryJob{Job: job}
				}, func(job model.CustomNamespaceJob) {
					jobToRun = cloudwatchrunner.CustomNamespaceJob{Job: job}
				},
			)
```
I might need to pass through a new `ResourceProvider` type but ideally can mask it through the ResourceEnrichment pipeline that will be setup.